Api stubs

richardkiss opened the pull request:
Create a smaller version of `XApi` as `XApiStub` to remove cycles from `chia.apis` <=> `chia.X` for X in `full_node`, `harvester`, `farmer`, etc.

richardkiss committed:
harvester_api_protocol

richardkiss committed:
move to apis

richardkiss committed:
Remove `chia.apis` => `chia.harvester` dependency.

richardkiss committed:
Remove circular dependencies between `chia.apis` and service modules.

richardkiss committed:
SolverAPIStub, and add solver to tach.toml.

Quexington reviewed (commented):
I'm concerned about the maintainability of this paradigm.  Is there anything enforcing consistency between these stubs and the actual objects?  I'm also a little confused why this... works?  The `@metadata.request` wrappers are supposed to wrap the actual methods.  Why does it work on these unimplemented stubs?  CI is passing I guess but it doesn't make sense to me.

Have you considered addressing this by simply removing the `APIProtocolRegistry` import from all of the individual packages?  It seems like it's entirely imported in every package to just pass directly to the `Service` class:
```
from chia.apis import APIProtocolRegistry
...
return Service(
...
class_for_type=APIProtocolRegistry
)
```
you could just have the `Service` object import the `APIProtocolRegistry` on its own and maybe leave that initialization argument for overrides (that no service will take advantage of at the present moment)

Another potential solution that might be more involved is to move the `start_*_service` files into a more root level package.  Conceptually, there's nothing wrong with these imports, these functions, since they're entry points, belong at a root-most level that imports from almost everything.  The libraries they are a part of really just need to provide the python APIs and wrapping that with a web service interface belongs in a conceptual package that imports from the libraries.

prozacchiwawa commented:
quexington's suggestion might be better than mine, but this might be an alternate idea that would help unify these big types if removing the stubs isn't desired.  the purpose is to relate the original provider types with the new classes provided in the stub with a single-direction dependency.  one possibly desired side effect is that the new protocols also separately describe the interfaces to the original api providers without the code inside them.

*snip*

Would it make sense to add prototype protocols in ./chia/apis and make both these exports and the real exports conform to them?  It seems that you can do this pretty simply when boiled down.

I feel like part of the unease on this might be that the new types added in stubs aren't connected to the main ones and so could drift.

For example, to make each base class a Protocol so mypy wants to check membership:

    class FullNodeApiProtocol(Protocol):
        ...
    
Then define a class that inherits it (it actually has no abstract methods):

    class FullNodeApiStub(FullNodeApiProtocol):
        if TYPE_CHECKING:
            from chia.server.api_protocol import ApiProtocol
    
            _protocol_check: ClassVar[ApiProtocol] = cast("FullNodeApiStub", None)

And make the original implement the protocol version:

    class FullNodeAPI:
        if TYPE_CHECKING:
            from chia.apis.full_node_stub import FullNodeApiProtocol
    
            _protocol_check: ClassVar[FullNodeApiProtocol] = cast("FullNodeAPI", None)

Then it'd be true that the code is getting more strict instead of having something that could potentially diverge invisibly.

What do you think?

richardkiss commented:
> I'm concerned about the maintainability of this paradigm. Is there anything enforcing consistency between these stubs and the actual objects? I'm also a little confused why this... works? The `@metadata.request` wrappers are supposed to wrap the actual methods. Why does it work on these unimplemented stubs? CI is passing I guess but it doesn't make sense to me.

I have to admit I don't completely understand all the horrible things `@metadata.request` does. Its complexity attempts to deal with the careless protocol that we have settled on. I have a lot of ideas about how a good protocol would work that I won't get into here. But if you want type-checking on RPC calls (the general meaning of RPC, not what chia calls "RPC"), even the best, simplest protocol must have stubs of some form. More on this below.

> Have you considered addressing this by simply removing the `APIProtocolRegistry` import from all of the individual packages? It seems like it's entirely imported in every package to just pass directly to the `Service` class:
> 
> ```
> from chia.apis import APIProtocolRegistry
> ...
> return Service(
> ...
> class_for_type=APIProtocolRegistry
> )
> ```
> 
> you could just have the `Service` object import the `APIProtocolRegistry` on its own and maybe leave that initialization argument for overrides (that no service will take advantage of at the present moment)

This would mean `chia.service` imports `chia.apis` which imports all the service API types (like `FullNodeAPI` and `WalletAPI`). But `start_wallet.py` uses `chia.service`. So we still have wallet depending on full node and vice versa.

> Another potential solution that might be more involved is to move the `start_*_service` files into a more root level package. Conceptually, there's nothing wrong with these imports, these functions, since they're entry points, belong at a root-most level that imports from almost everything. The libraries they are a part of really just need to provide the python APIs and wrapping that with a web service interface belongs in a conceptual package that imports from the libraries.

I actually fairly recently moved the `start_xxx.py` scripts to their respective service module. I don't remember if I considered or tried moving them one level higher. But yes, bumping these scripts higher up in the dependency hierarchy would resolve this particular dependency loop I'm trying to break in `tach.toml`.

At the same time, I still think it's not sufficient to do this. Consider the thought experiment where we split the full node and the wallet into separate repos. Where would `start_wallet.py` and `start_full_node.py` live? They would have to live at a repo above both `chia.full_node` and `chia.wallet`. In other words, you would not be able to run a wallet without including the code for `chia.full_node`.

Fundamentally, we have a problem common to many client/server systems: the client wants to be able to call a server implementation and have those calls type-checked. Consider a server and client:

```python
class Server:
    async def triple(self, v: int) -> int:
         return v * 3
```

```python
# client

async def main():
    server = await find_server()
    print(await server.triple(10))
```


What is the type of `server` in the client code? The idea is that on the client side, `server` is a proxy object with methods that exactly match the remote server side, that, when called, it bundles up the arguments into an RPC message, invokes the server, gets the response, unbundles the response into a return value, and returns it.

This way get type checking on our calls. But unless we include the `server` code in the `client` project, we have to create a `stub` version that includes declarations but not implementations.

We are *sort of* doing this in chia; in particular, the (oddly named) `@metadata.request` decorator looks at the parameters and at runtime generates code to turn invocations into messages that are sent. I think we might actually be able to remove these decorators from the implementations; it should be enough to have them on the stubs. But I don't completely understand it. It seems like it's wrapping stuff for both the client AND the server. It would be nice to crack this open and simplify it a bit so a casual observer has any hope of understanding it, but that's wildly out of scope here.

richardkiss committed:
Make the stubs (lack of) implementations more consistent.

richardkiss commented:
> quexington's suggestion might be better than mine, but this might be an alternate idea that would help unify these big types if removing the stubs isn't desired. the purpose is to relate the original provider types with the new classes provided in the stub with a single-direction dependency. one possibly desired side effect is that the new protocols also separately describe the interfaces to the original api providers without the code inside them.
> 
> _snip_
> 
> Would it make sense to add prototype protocols in ./chia/apis and make both these exports and the real exports conform to them? It seems that you can do this pretty simply when boiled down.
> 
> I feel like part of the unease on this might be that the new types added in stubs aren't connected to the main ones and so could drift.
> 
> For example, to make each base class a Protocol so mypy wants to check membership:
> 
> ```
> class FullNodeApiProtocol(Protocol):
>     ...
> ```
> 
> Then define a class that inherits it (it actually has no abstract methods):
> 
> ```
> class FullNodeApiStub(FullNodeApiProtocol):
>     if TYPE_CHECKING:
>         from chia.server.api_protocol import ApiProtocol
> 
>         _protocol_check: ClassVar[ApiProtocol] = cast("FullNodeApiStub", None)
> ```
> 
> And make the original implement the protocol version:
> 
> ```
> class FullNodeAPI:
>     if TYPE_CHECKING:
>         from chia.apis.full_node_stub import FullNodeApiProtocol
> 
>         _protocol_check: ClassVar[FullNodeApiProtocol] = cast("FullNodeAPI", None)
> ```
> 
> Then it'd be true that the code is getting more strict instead of having something that could potentially diverge invisibly.
> 
> What do you think?

I thought this was being done already, but I guess it isn't. There are a couple simple lines I can add that will have mypy ensure that `FullNodeApi` implements `FullNodeApiStub`, and so on.

For some reason, python has two ways to do this: `Protocol` https://typing.python.org/en/latest/spec/protocol.html and `abc` https://docs.python.org/3/library/abc.html ; I'll try making the stubs inherit from `Protocol`.

richardkiss committed:
Use `stub_metadata_for_type` instead of `class_for_type`.

The only thing ever used from `ApiProtocol` in `WSConnection`
was the metadata field, so let's pass along only that.

richardkiss committed:
NewSignagePointHarvester2

prozacchiwawa reviewed (dismissed):
neat

richardkiss committed:
remove some stuff
